src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -32,73 +33,414 @@
+  for (size_t idx = 0; idx < _max; idx++) {
+    _membership[idx] = NotFree;
+  }
+
+  for (size_t partition_id = 0; partition_id < NumPartitions; partition_id++) {
+    _leftmosts[partition_id] = _max;
+    _rightmosts[partition_id] = 0;
+    _leftmosts_empty[partition_id] = _max;
+    _rightmosts_empty[partition_id] = 0;
+    _capacity[partition_id] = 0;
+    _used[partition_id] = 0;
+  }
+
+  _region_counts[Mutator] = _region_counts[Collector] = 0;
+  _region_counts[NotFree] = _max;
+}
+
+void ShenandoahRegionPartition::increase_used(ShenandoahFreeSetPartitionId which_partition, size_t bytes) {
+  assert (which_partition > NotFree && which_partition < NumPartitions, "Partition must be valid");
+  _used[which_partition] += bytes;
+  assert (_used[which_partition] <= _capacity[which_partition],
+          "Must not use (" SIZE_FORMAT ") more than capacity (" SIZE_FORMAT ") after increase by " SIZE_FORMAT,
+          _used[which_partition], _capacity[which_partition], bytes);
+}
+
+inline void ShenandoahRegionPartition::shrink_interval_if_boundary_modified(ShenandoahFreeSetPartitionId partition, size_t idx) {
+  if (idx == _leftmosts[partition]) {
+    while ((_leftmosts[partition] < _max) && !in_partition(_leftmosts[partition], partition)) {
+      _leftmosts[partition]++;
+    }
+    if (_leftmosts_empty[partition] < _leftmosts[partition]) {
+      // This gets us closer to where we need to be; we'll scan further when leftmosts_empty is requested.
+      _leftmosts_empty[partition] = _leftmosts[partition];
+    }
+  }
+  if (idx == _rightmosts[partition]) {
+    while (_rightmosts[partition] > 0 && !in_partition(_rightmosts[partition], partition)) {
+      _rightmosts[partition]--;
+    }
+    if (_rightmosts_empty[partition] > _rightmosts[partition]) {
+      // This gets us closer to where we need to be; we'll scan further when rightmosts_empty is requested.
+      _rightmosts_empty[partition] = _rightmosts[partition];
+    }
+  }
+}
+
+inline void ShenandoahRegionPartition::expand_interval_if_boundary_modified(ShenandoahFreeSetPartitionId partition,
+                                                           size_t idx, size_t region_available) {
+  if (region_available == _region_size_bytes) {
+    if (_leftmosts_empty[partition] > idx) {
+      _leftmosts_empty[partition] = idx;
+    }
+    if (_rightmosts_empty[partition] < idx) {
+      _rightmosts_empty[partition] = idx;
+    }
+  }
+  if (_leftmosts[partition] > idx) {
+    _leftmosts[partition] = idx;
+  }
+  if (_rightmosts[partition] < idx) {
+    _rightmosts[partition] = idx;
+  }
+}
+
+// Remove this region from its free partition, but leave its capacity and used as part of the original free partition's totals.
+// When retiring a region, add any remnant of available memory within the region to the used total for the original free partition.
+void ShenandoahRegionPartition::retire_within_partition(size_t idx, size_t used_bytes) {
```

Comment:
Why is the method called `retire_within_partition()` instead of `retire_from_partition()` ?

(i.e. why _within_ partition, since it's leaving its free partition?)

Reply:
It's a little subtle.  Maybe it needs more documentation.  We retire the region so it no longer is within the range searched when new allocations are made.  However, its totals (capacity and used) are still counted toward the Mutator or Collector partition's total.

We've probably created some of our own pain by calling this a partition instead of a "free set".  Advice?

Reply:
I'll change the name, as you suggest.